rfd/0159-password-status.md

```diff
@@ -0,0 +1,125 @@
+---
+authors: Bartosz Leper (bartosz.leper@goteleport.com)
+state: draft
+---
+
+# RFD 0159 - Password Status
+
+## Required Approvers
+
+- Engineering: (@zmb3 || @codingllama)
+- Security: @jentfoo
+- Product: @roraback
+
+## What
+
+We want to improve the UX of our flows related to changing passwords and management of authentication devices (both MFA and passwordless). The following new features must be supported:
+
+- The user must be able to set password having only passwordless credentials configured.
+- We should do our best to indicate which authentication methods a user has configured. Unfortunately, due to past design decisions, we will not be able to do it in every case; see the [Details](#details) section.
+
+## Why
+
+We are promoting passwordless authentication and would like to make it a default in our cloud environments. We need to decrease the UX friction and fill the missing gaps.
+
+Related issues:
+
+- [Design for defaulting to "Passwordless" authentication + updating view to clarify options](https://github.com/gravitational/cloud/issues/5757)
+- [Remove local password authentication from teleport](https://github.com/gravitational/teleport/issues/19671)
+
+## Details
+
+For the purpose of this document, we define "passwordless key/token" as a resident WebAuthn key.
+
+Currently, it's impossible to tell whether a user has a password configured. This is because the user always has _some_ password set. If they never set it up themselves, we just use a random 16-character password. As a consequence, when the user wants to set a password, we always ask them for the current one. It makes it impossible to add a password to account that already has a passwordless key configured and doesn't have any current password.
```

Comment:
Could we instead not set a password instead of a random password? Now, the issue of dealing with timing attacks would likely come up, but we already have some mitigations in place to consider this concern when a user doesn't exist. Look in `lib/auth/password.go` at `checkPasswordWOToken()` and note how it uses `fakePasswordHash` to still run a password validity check even though there is no user or password at all. How hard would it be to do the same or similar thing when there is no password?

Reply:
I confess I had similar thoughts, not setting the password seems more intuitive to me. No strong opinions though.

Reply:
@reedloden If we implemented a "no password, no password hash" strategy, we still would like to implement the password status flag. That's because yes, for _new_ users, we could simply not set a password hash if they don't have one or set a password hash if they do, but for the existing users, we would have no way of distinguishing a real hash from a random one.

The reason why I didn't want to make this change is that I considered the value of it too low to justify changing a mission-critical and security-critical code. However, right now when I look at it, it looks the "hash not found" is actually already treated by the authentication code ([checkPasswordWOToken](https://github.com/gravitational/teleport/blob/c26314652afbea1ecddaf1462c94988f6ceb16b4/lib/auth/password.go#L171)) as "user not found" and leads to - guess what - exactly the fake hash comparison path. So I guess that changes to the authentication code path will be limited to changing the debug message. I can live with that, so I think I'll use your suggestion.

Again, that doesn't mean we don't need the status field, but yes, we can incidentally simplify this part.